drivers/hmis/app/models/hmis/entity_access_loader_factory.rb

```diff
@@ -0,0 +1,148 @@
+##
+
+# Factory to resolve the user access loader for a given entity. It also resolves
+# the entity that should be passed to the loader. A resolver blocker is used to
+# traverse associations as, in the case of graphql, another data loader would be
+# used to resolve the association.
+#
+# It's up to the caller to actually invoke the loader with the resolved entity to
+# perform the access check.
+#
+# @example
+#   factory = Hmis::EntityAccessLoaderFactory.new(user)
+#   loader, entity = factory.perform(project) do |record, association|
+#     # resolve an entity association. In graphql, a data loader would be used
+#     association.send(record)
+#   end
+#   # resolve a user's permission on project through the loader. Loader.fetch() would
+#   # be used to called for many projects
+#   allowed = loader.fetch_one(entity, :can_do_something)
+class Hmis::EntityAccessLoaderFactory
+  # @param user [Hmis::User]
+  def initialize(user)
+    @user = user
+  end
+
+  # Given an entity, return a tuple of the loader and the entity to check against
+  #
+  # @param entity [#entity_record] active record entity
+  # @yieldparam [#entity_record] record
+  # @yieldparam [Symbol] association name
+  # @yieldreturn [#resolved, nil] the association (record.association)
+  # @return [Array<Hmis::BaseLoader, #entity>]
+  def perform(entity, &block)
+    if entity.new_record?
+      resolve_new_record(entity, safety: 0, &block)
+    else
+      resolve_entity(entity, safety: 0, &block)
+    end
+  end
+
+  protected
+
+  def resolve_entity(entity, safety:, &block)
+    check_safety(safety)
+
+    loader = case entity
+    when Hmis::Hud::Client
+      Hmis::Hud::ClientAccessLoader
+    when Hmis::Hud::Project
+      Hmis::Hud::ProjectAccessLoader
+    when Hmis::Hud::Organization
+      Hmis::Hud::OrganizationAccessLoader
+    when GrdaWarehouse::DataSource
+      Hmis::DataSourceAccessLoader
+    end
+    if loader
+      [loader.new(@user), entity]
+    else
+      # recurse
+      resolve_association(entity, safety: safety + 1, &block)
+    end
+  end
+
+  # follow the association chain for entities that don't have their own data loader
+  def resolve_association(entity, safety:, &block)
+    check_safety(safety)
+
+    resolved = case entity
+    when Hmis::File
+      # Files are always linked to a client, and optionally link to a specific
+      # enrollment. If the file is linked to an enrollment, access to the file
+      # should be limited based on access to that enrollment.
+      entity.enrollment_id ? block.call(entity, :enrollment) : block.call(entity, :client)
+    when Hmis::Hud::Enrollment
+      if entity.in_progress?
+        block.call(entity, :wip)
+      else
+        block.call(entity, :project)
+      end
+    when Hmis::Hud::HmisService, Hmis::Hud::Service, Hmis::Hud::CustomService
+      # optimization for persisted records only
+      block.call(entity, :project)
```

Comment:
What is this giving us? It looks like the fallthrough case `resolve_through_project` does the exact same thing 🧐

Reply:
Oh oops! This is some residue from a refactor. I removed this case, thanks for spotting!